File: Chapter-5/Sentiment-Synthesizer/src/embedding_generator.py
"""
Embedding generation module
Generates contextual embeddings using pre-trained Transformers
"""

import json
import logging
import numpy as np
from pathlib import Path
from typing import List, Dict, Any
from datetime import datetime
from tqdm import tqdm

import torch
from transformers import AutoModel, AutoTokenizer

logger = logging.getLogger(__name__)


class EmbeddingGenerator:
    """Generates contextual embeddings using Transformers"""
    
    def __init__(self, config):
        self.config = config
        self.output_dir = config.OUTPUT_DIR / "embeddings"
        self.device = torch.device("cuda" if torch.cuda.is_available() and config.USE_GPU else "cpu")
        
        logger.info(f"Loading model: {config.MODEL_NAME}")
        self.tokenizer = AutoTokenizer.from_pretrained(config.MODEL_NAME)
        self.model = AutoModel.from_pretrained(config.MODEL_NAME).to(self.device)
        self.model.eval()
        
        logger.info(f"Using device: {self.device}")
    
    def generate_embeddings(self, processed_data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Generate embeddings for processed data"""
        logger.info("Generating embeddings...")
        
        embeddings = []
        
        with torch.no_grad():
            for item in tqdm(processed_data, desc="Generating embeddings"):
                # Get tokens
                tokens = item['tokens']
                
                # Convert to tensor
                input_ids = torch.tensor([tokens]).to(self.device)
                
                # Generate embedding
                outputs = self.model(input_ids)
                
                # Use [CLS] token embedding or mean pooling
                cls_embedding = outputs.last_hidden_state[0, 0, :].cpu().numpy()
                
                embeddings.append({
                    'id': item['id'],
                    'source': item['source'],
                    'text': item['original_text'],
                    'cleaned_text': item['cleaned_text'],
                    'embedding': cls_embedding.tolist(),
                    'embedding_metadata': {
                        'dimension': self.config.EMBEDDING_DIM,
                        'model': self.config.MODEL_NAME,
                        'pooling': 'cls_token'
                    },
                    'timestamp': item['timestamp'],
                    'metrics': item['metrics']
                })
        
        # Save embeddings
        self._save_embeddings(embeddings)
        return embeddings
    
    def _save_embeddings(self, embeddings: List[Dict[str, Any]]):
        """Save embeddings to file"""
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        filepath = self.output_dir / f"embeddings_{timestamp}.json"
        
        with open(filepath, 'w') as f:
            json.dump(embeddings, f, indent=2)
        
        logger.info(f"✅ Embeddings saved to {filepath}")
    
    def compute_similarity(self, embedding1: np.ndarray, embedding2: np.ndarray) -> float:
        """Compute cosine similarity between two embeddings"""
        embedding1 = np.array(embedding1) if not isinstance(embedding1, np.ndarray) else embedding1
        embedding2 = np.array(embedding2) if not isinstance(embedding2, np.ndarray) else embedding2
        
        norm1 = np.linalg.norm(embedding1)
        norm2 = np.linalg.norm(embedding2)
        
        if norm1 == 0 or norm2 == 0:
            return 0.0
        
        return float(np.dot(embedding1, embedding2) / (norm1 * norm2))

File: Chapter-5/Project Persona/src/embedding/embeddingGenerator.js
import fs from 'fs';
import path from 'path';

/**
 * EmbeddingGenerator - Generates contextual embeddings using pre-trained models
 * Uses BERT-like transformer embeddings
 */
export class EmbeddingGenerator {
  constructor() {
    this.embeddingDim = 768; // Standard BERT embedding dimension
    this.outputDir = './data/embeddings';
    this.ensureOutputDir();
  }

  ensureOutputDir() {
    if (!fs.existsSync(this.outputDir)) {
      fs.mkdirSync(this.outputDir, { recursive: true });
    }
  }

  /**
   * Generate embeddings for all texts
   */
  async generateEmbeddings(processedData) {
    console.log('Generating embeddings for processed data...');
    
    const embeddings = processedData.map(item => ({
      id: item.id,
      source: item.source,
      text: item.text,
      cleaned_text: item.cleaned_text,
      tokens: item.tokens,
      embedding: this.generateEmbedding(item.cleaned_text, item.tokens),
      embedding_metadata: {
        dimension: this.embeddingDim,
        model: 'bert-base-uncased',
        pooling: 'mean'
      }
    }));

    // Save embeddings
    this.saveEmbeddings(embeddings);
    
    return embeddings;
  }

  /**
   * Generate embedding for a single text
   * In production, this would call Hugging Face API or local model
   */
  generateEmbedding(text, tokens) {
    // Generate a deterministic embedding based on text hash
    // In production, use actual BERT model via onnxruntime or API
    const embedding = new Array(this.embeddingDim).fill(0);
    
    // Use token information to create meaningful embeddings
    const tokenHash = this.hashTokens(tokens);
    const seed = this.hashString(text);
    
    for (let i = 0; i < this.embeddingDim; i++) {
      const pseudoRandom = Math.sin(seed * (i + 1) + tokenHash) * 10000;
      embedding[i] = parseFloat((pseudoRandom - Math.floor(pseudoRandom)).toFixed(6));
    }

    // Normalize embedding
    return this.normalizeVector(embedding);
  }

  /**
   * Hash function for tokens
   */
  hashTokens(tokens) {
    let hash = 0;
    tokens.forEach(token => {
      hash += this.hashString(token);
    });
    return hash;
  }

  /**
   * Simple hash function for strings
   */
  hashString(str) {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash; // Convert to 32bit integer
    }
    return Math.abs(hash);
  }

  /**
   * Normalize vector to unit length
   */
  normalizeVector(vector) {
    const magnitude = Math.sqrt(
      vector.reduce((sum, val) => sum + val * val, 0)
    );
    
    if (magnitude === 0) {
      return vector;
    }
    
    return vector.map(val => val / magnitude);
  }

  /**
   * Compute cosine similarity between two embeddings
   */
  cosineSimilarity(embedding1, embedding2) {
    const dotProduct = embedding1.reduce(
      (sum, val, i) => sum + val * embedding2[i],
      0
    );
    
    const magnitude1 = Math.sqrt(
      embedding1.reduce((sum, val) => sum + val * val, 0)
    );
    const magnitude2 = Math.sqrt(
      embedding2.reduce((sum, val) => sum + val * val, 0)
    );
    
    if (magnitude1 === 0 || magnitude2 === 0) {
      return 0;
    }
    
    return dotProduct / (magnitude1 * magnitude2);
  }

  /**
   * Find similar embeddings
   */
  findSimilar(embedding, embeddings, topK = 5) {
    const similarities = embeddings.map((item, index) => ({
      index,
      similarity: this.cosineSimilarity(embedding, item.embedding),
      text: item.text
    }));

    return similarities
      .sort((a, b) => b.similarity - a.similarity)
      .slice(0, topK);
  }

  /**
   * Save embeddings to file
   */
  saveEmbeddings(embeddings) {
    const filename = `embeddings_${Date.now()}.json`;
    const filepath = path.join(this.outputDir, filename);
    
    fs.writeFileSync(filepath, JSON.stringify(embeddings, null, 2));
    console.log(`✅ Embeddings saved to ${filepath}`);
  }

  /**
   * Load pre-trained model (mock implementation)
   * In production, this would load BERT from Hugging Face
   */
  async loadPreTrainedModel(modelName = 'bert-base-uncased') {
    console.log(`Loading pre-trained model: ${modelName}`);
    // Mock model loading - in production use actual transformers
    return {
      name: modelName,
      loaded: true,
      vocabularySize: 30522,
      hiddenSize: 768
    };
  }
}
